src/compiler/checker.ts

```diff
@@ -6924,20 +6946,13 @@ namespace ts {
             let signatureList: Signature[];
             let types = (<UnionType>type).types;
             for (let current of types) {
-                // The signature set of all constituent type with call signatures should match
-                // So number of signatures allowed is either 0 or 1
-                if (signatureList &&
-                    getSignaturesOfStructuredType(current, SignatureKind.Call).length > 1) {
-                    return undefined;
-                }
-
                 let signature = getNonGenericSignature(current);
                 if (signature) {
                     if (!signatureList) {
                         // This signature will contribute to contextual union signature
                         signatureList = [signature];
                     }
-                    else if (!compareSignatures(signatureList[0], signature, /*compareReturnTypes*/ false, compareTypes)) {
+                    else if (!compareSignatures(signatureList[0], signature, /*partialMatch*/ false, /*ignoreReturnTypes*/ true, compareTypes)) {

```

Comment:
It seems like in both of those examples, you'd get no contextual signature, whether or not you accept partial matches. But something like the following would give you a contextual signature *only* if you accept partial matches:
```ts
type RequiredCallable = (s: string) => string;
type OptionalCallable = (s?: string) => string;
var func: RequiredCallable | OptionalCallable = x => x.charAt(0); // x is string only if partialMatches is true
```
It's a pretty silly example, but my point is that partialMatches will only make it more likely that a contextual signature will emerge at all.